src/ukify/ukify.py

```diff
@@ -0,0 +1,592 @@
+            raise ValueError(f'Cannot find {cls.TEXT_PATTERN!r} in {filename}')
+
+        return m.group(1).decode()
+
+    @classmethod
+    def scrape(cls, filename, opts=None):
+        for func in (cls.scrape_x86, cls.scrape_elf, cls.scrape_generic):
+            try:
+                version = func(filename, opts=opts)
+                print(f'Found uname version: {version}')
+                return version
+            except Exception as e:
+                print(str(e))
+        return None
+
+
+@dataclasses.dataclass
+class Section:
+    name: str
+    content: pathlib.Path
+    tmpfile: typing.Optional[typing.IO] = None
+    offset: typing.Optional[int] = None
+    measure: bool = False
+
+    @classmethod
+    def create(cls, name, contents, measure=False):
+        if isinstance(contents, str):
+            tmp = tempfile.NamedTemporaryFile(mode='wt', prefix=f'tmp{name}')
+            tmp.write(contents)
+            tmp.flush()
+            contents = pathlib.Path(tmp.name)
+        else:
+            tmp = None
+
+        return cls(name, contents, tmpfile=tmp, measure=measure)
+
+    @classmethod
+    def parse_arg(cls, s):
+        try:
+            name, contents, *rest = s.split(':')
+        except ValueError as e:
+            raise ValueError(f'Cannot parse section spec (name or contents missing): {s!r}') from e
+        if rest:
+            raise ValueError(f'Cannot parse section spec (extraneous parameters): {s!r}')
+
+        if contents.startswith('@'):
+            contents = pathlib.Path(contents[1:])
+
+        return cls.create(name, contents)
+
+    def size(self):
+        return self.content.stat().st_size
+
+    def check_name(self):
+        # PE section names with more than 8 characters are legal, but our stub does
+        # not support them. A dot is prepended, so the name must be at most 7 chars.
+        name = f'.{self.name}'
+        if not name.isascii() or not name.isprintable():
+            raise ValueError(f'Bad section name: {name!r}')
+        if len(name) > 8:
+            raise ValueError(f'Section name too long: {name!r}')
+
+
+@dataclasses.dataclass
+class UKI:
+    executable: [pathlib.Path|str]
+    sections: list[Section] = dataclasses.field(default_factory=list, init=False)
+    offset: int = dataclasses.field(default=None, init=False)
+
+    def __post_init__(self):
+        self.offset = round_to_page(pe_executable_size(self.executable))
+
+    def add_section(self, section):
+        assert self.offset
+        assert section.offset is None
+        section.offset = self.offset
+        self.offset += round_to_page(section.size())
+        self.sections += [section]
+
+
+def parse_banks(s):
+    banks = s.split(',')
+    # TODO: do some sanity checking here
+    return banks
+
+
+def check_splash(filename):
+    if filename is None:
+        return
+
+    # import is delayed, to avoid import when the splash image is not used
+    try:
+        from PIL import Image
```

Comment:
Isn't PIL python2? Afaik, Pillow is the go-to now.

Reply:
The fork Pillow took over `PIL` as module, the same import is [used in their tutorial](https://pillow.readthedocs.io/en/stable/handbook/tutorial.html).

Reply:
Yeah, `import PIL` means `pillow` for the last few years.

Reply:
I knew there is a reason why I don't do python…